src/client.ts

```diff
@@ -86,6 +86,11 @@ export class LuneClient {
     public setAccount(accountId: string) {
         this.config.ACCOUNT = accountId
     }
+
+    // TODO generate this via the base generator. This was manually added to showcase.
+    public setApiVersion(apiVersion: 'v1' | 'v2' | 'v3') {
```

Comment:
This needs to be changed.

Stripe only allows one type of version.

Reply:
Stripe still allows you to manually override the version, but the types involved are unknown so you have to manually cast things. Do you think this method should just receive a string, not exist at all and have calendar version customisable only on a per endpoint basis or...?

Reply:
Okay, I was still in holiday mode so let me explore this problem a bit further and give a bit more information on why I believe having versions here is helpful. First a few notes:
- Each `lune-ts` client is meant to be tied to a specific version. So the types on each method will only refer to the latest version.
- Each endpoint/method in the client will have a new property that allows to override the version that is meant to be called. Since the types only refer to the latest one the user will have to manually perform type casts if the method signature has changed.
- Whenever a new API calendar version is created a new `lune-ts` version should accompany it.
- A new `lune-ts` version might not mean changes in the API schema.

So, the usual workflow in our side will be very similar. With the addition of API versioning, our customers will now need to be a bit more aware when upgrading the library since upgrading to a new version might require changes, but the changelog should guide them and types should catch any issues.

So I see two main questions:
1. Which versions do we want to explicitly support in a `lune-ts` release
2. Do we want to allow changing the version only on a per endpoint or for the whole client

For 1. while true that a specific release has types tied to a specific release, we definitely want to allow clients to use old versions so they're able to upgrade the library and keep their old version. This can come very much in handy when changes to the actual library are done (imagine new options, performance improvements, bug fixes etc.). Adding them as types just makes everything type safety. Keep in mind that the user can manual type cast and use whatever value it wants for version, even future ones if for some reason they want to keep the old client but use the new API version.

For 2. while true that setting the version for the whole client isn't a requirement, it makes things much easier for our customers. If they want to use a specific version they only need one change, instead of having to adapt all their library calls.

Reply:
> Do we want to allow changing the version only on a per endpoint or for the whole client
I am for this because of the simplicity of dealing with one version only.

Reply:
> I am for this because of the simplicity of dealing with one version only.

What are you referring when you mean this here? Only allowing changing version on a per endpoint basis? If we give them the option to override a specific endpoint why not an option to do it for all of them at once? If you allow on a per endpoint basis but not globally you're simply making their life harder, I don't really see a point for that.

If we really follow this logic of only having one version then we will force them to upgrade/downgrade their code to a single version. It feels to me that this will just make users never upgrade and when they because they want something it will force them to upgrade other things they don't really care about.

Reply:
Apologies for the miscommunication.
I prefer one version for the entire client, not per endpoint version.

Reply:
I see...

But if we only allow setting the version for the entire client then this means that a customer that wants to have new functionalities that were introduced in an endpoint will be forced to make changes for all endpoints that changed in the meantime, even if they're not interested in them 🤔 

I think giving flexibility and having both options is important because a library update might happen for multiple reasons and our costumers might want to use new features while keeping their changes minimal, which means different strategies are more appropriate depending on the changes introduced.
- If a small amount of breaking changes happened in the API the client will probably:
    - Update it's code to use the latest version. This might not be possible for them
    - Update the library to the latest version and if some endpoint proves hard to do this simply override the version on that endpoint and use the previous one.
- If a lot of breaking changes happened in the API:
    - Update it's code to use the latest version.  This might not be possible for them
    - Update the library and override it to use the previous version. This will require type casting on affected endpoints. There's no point in doing this unless there are library improvements that are not tied to the API.
    - Update the library and mix mash endpoints, on some they might want to upgrade if changes are small but on others they might not if it's impossible. This allows library improvements to be assimilated while keeping a subset of endpoints endpoints using a previous version.

I think if we don't have per endpoint version then we will be prohibiting a lot of customers to upgrade versions if for example a new parameter they don't have is now required for an endpoint (even an endpoint they don't care much about).

At the end of the day though, this is up to us and how we want the library to operate. But I find it weird that the API allows to set a specific API version on an endpoint but using our client it's impossible.

Reply:
yes, I understand.

I think it's just easier to think one version at the time.

My opinion here is 'soft'

Reply:
I agree with that. Hence why a library will have a specific version attached to it. But on some cases it might not be enough to make life easier to our customers I think.

I also don't particularly have strong opinions here. But overall do think giving flexibility to the user is probably the best approach.

Reply:
@jstasiak I know this is a big discussion but do you have any thoughts on this? The question is if we should allow on `lune-ts` to specify the API version on a per endpoint basis only or also have a way to set the API version that will be used for all subsequent calls.

Reply:
All right, I went through the comments and I think I'm up to speed on this.

https://lune-climate.slack.com/archives/C02231YFEMP/p1725880222437859 is an internal discussion where I educated myself on the general approach to delivering this feature.

It's my opinion that *regardless* of the discussion above (version selection per endpoint, globally or some mix of those) the solution cannot decrease the degree of type safety in any way.

"cannot decrease the degree of type safety" means that when I call `createClientAccount` with parameters of type `P` and I get a return value of type `R` in return I want the following to be true and verified by the compiler: the version of the `createClientAccount` API endpoint that's ultimately gonna be used by the client takes inputs of type `P` and returns `R` on success.

I would discard any change to the client that violates this assumption. I can't currently imagine any potential upside of this feature that would be worth this price (decreased type safety) to me. That means:

1. Client-global API version switch – can't be made type-safe, I'd say off the table.
2. Per-endpoint API version switch – these seem all right if we can provide type-safe versioned function overloads, "no deal" otherwise.

Reply:
Regarding the above, changes in the base generator to accommodate have been done on the PR that already existed and a rereview requested. See: https://github.com/lune-climate/openapi-typescript-codegen/pull/73

I'll modify this PR to add the changes in `lune-ts` once that is merged and a new version of the base generator is available.

Reply:
FWIW it's not my intention to block the effort if you want to pursue that direction, it's just that I was asked for an opinion (I admit it's a strong one).